File: deepspark/spark16/src/main/java/sparklearn/rdd/transformations/LearnaggregateByKey2.java
/** 
 * Project Name:sparklearn 
 * File Name:LearnaggregateByKey2.java 
 * Package Name:sparklearn.rdd.transformations 
 * Date:2017年7月6日下午4:21:35 
 * sky.zyl@hotmail.com
*/  
  
package sparklearn.rdd.transformations;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;

import scala.Tuple2;
import sparklearn.util.PrintUtilPro;

/** 
 * ClassName:LearnaggregateByKey2 <br/> 
 * Function: TODO ADD FUNCTION. <br/> 
 * Reason:   TODO ADD REASON. <br/> 
 * Date:     2017年7月6日 下午4:21:35 <br/> 
 * @author   hadoop 
 * @version   
 * @see       
 */
public class LearnaggregateByKey2 {

	public static void main(String[] args) {
		SparkConf conf = new SparkConf().setAppName(LearnaggregateByKey2.class.getSimpleName()).setMaster("local");

		JavaSparkContext sc = new JavaSparkContext(conf);
		
		List<Tuple2<Integer, Integer>> datas = new ArrayList<>();
	    datas.add(new Tuple2<>(1, 3));
	    datas.add(new Tuple2<>(1, 2));
	    datas.add(new Tuple2<>(1, 4));
	    datas.add(new Tuple2<>(2, 3));

	    JavaPairRDD<Integer, Integer> strRDD = sc.parallelizePairs(datas, 2)
	            .aggregateByKey(
	                    0,
	                    new Function2<Integer, Integer, Integer>() {
	                        @Override
	                        public Integer call(Integer v1, Integer v2) throws Exception {
	                            System.out.println("seq: " + v1 + "\t" + v2);
	                            return Math.max(v1, v2);
	                        }
	                    },
	                    new Function2<Integer, Integer, Integer>() {
	                        @Override
	                        public Integer call(Integer v1, Integer v2) throws Exception {
	                            System.out.println("comb: " + v1 + "\t" + v2);
	                            return v1 + v2;
	                        }
	                    });
	    //strRDD.collect().forEach(System.out::println);
	    PrintUtilPro.printList(strRDD.collect());
		
		

	}

}
 

File: deepspark/spark16/src/main/scala/sparklearnscala/rdd/transformations/LearnaggregateByKey2.scala
package sparklearnscala.rdd.transformations

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext

object LearnaggregateByKey2 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName(LearnaggregateByKey2.getClass.getSimpleName).setMaster("local")
    val sc = new SparkContext(conf)

    // 合并在同一个partition中的值，a的数据类型为zeroValue的数据类型，b的数据类型为原value的数据类型
    def seq(a:Int, b:Int): Int = {
        println("seq: " + a + "\t" + b)
        math.max(a, b)
    }

    // 合并在不同partition中的值，a,b的数据类型为zeroValue的数据类型
    def comb(a:Int, b:Int): Int = {
        println("comb: " + a + "\t" + b)
        a + b
    }

    //先在各自分区里内对value进行seq函数计算(zeroValue主要是为了在seq函数中作为第一个默认值来进行的，有一定的作用)，得到中间结果k v；然后把所有分区的value进行com函数操作
    
    // 数据拆分成两个分区
    // 分区一数据: (1,3) (1,2)
    // 分区二数据: (1,4) (2,3)
    // zeroValue 中位值，定义返回value的类型，并参与运算
    // seqOp 用来在一个partition中合并值的
    // 分区一相同key的数据进行合并
    // seq: 0   3  (1,3)开始和中位值合并为3
    // seq: 3   2  (1,2)再次合并为3
    // 分区二相同key的数据进行合并
    // seq: 0   4  (1,4)开始和中位值合并为4
    // seq: 0   3  (2,3)开始和中位值合并为3
    // comb 用来在不同partition中合并值的
    // 将两个分区的结果进行合并
    // key为1的, 两个分区都有, 合并为(1,7)
    // key为2的, 只有一个分区有, 不需要合并(2,3)
    sc.parallelize(List((1, 3), (1, 2), (1, 4), (2, 3)), 2)
        .aggregateByKey(0)(seq, comb)
        .collect()
        .foreach(println)
  }
}